### ISSUE TITLE
Publisher write call blocks or fails when a new subscriber is started later [7833]

### ISSUE THREAD

**Issue by aquad** [eProsima/Fast-RTPS#1042]
## Expected Behavior

The publisher's write call should not block excessively or fail when a new subscriber is started.

## Current Behavior

I have a publisher regularly writing e.g. at 100Hz, with reliable QoS, keep-last, depth 1.

Some minutes later, I start a subscriber to it. The publisher's write call then blocks for many milliseconds, sometimes over 100ms, resulting in failure. This only happens once, when the subscriber is started, after which it returns to normal.

## Steps to Reproduce

I have a simple ROS2 test publisher that keeps statistics of how long each publish call takes, and prints out when it's high.
https://github.com/aquad/test_pub_blocking_time

1. Create publisher running at 1000Hz:
   `ros2 run test_pub_blocking_time test_pub_blocking_time --rate 1000`
2. Wait 1 minute.
3. Create subscriber: `ros2 topic hz /test_topic`

With the above wait, it blocks for ~15ms. Waiting longer increases the blocking time.

## System information

- Fast-RTPS version: 1.8.0
- OS: Ubuntu 18.04
- Network interfaces: lo (127.0.0.1)
- ROS2: Dashing

## Diagnosis

Within `PublisherImpl::create_new_change_with_params()`, the mp_writer mutex is try-locked, which blocks, waiting on another thread. If it reaches 100ms, it will fail and return.

It is waiting on the mutex lock at `StatefulWriter::send_any_unsent_changes()`. Most of the time in this function is spent on handling a very large `std::set<SequenceNumber_t>` in notRelevantChanges (e.g.  size over 100000), around `RTPSMessageGroup::add_gap()`,

The set is created at `StatefulWriter::send_any_unsent_changes()`:
```
remoteReader->for_each_unsent_change(max_sequence, unsent_change_process);
```
The remoteReader (ReaderProxy) stores a `SequenceNumber_t changes_low_mark_` member, and it has to iterate from there to max_sequence (can be hundreds of thousands), adding elements to the std::set.  When you create a new subscriber, it results in a new ReaderProxy, with `changes_low_mark_ = 0`, and thus a giant std::set.

So the issue comes down to handling 'gap' submessages in the DDS RTPS spec. Gap submessages are meant to communicate the range/list of sequence numbers that are no longer relevant. However, if the minimum sequence number in the history is 100000 when the ReaderProxy is created, I don't think it needs to create a gap submessage saying messages 0-100000 are no longer valid. The DDS RTPS spec document seems to indicate that any sequence number to put in a gap message is triggered from the_reader_proxy.next_requested_change() or the_reader_proxy.next_unsent_change(). I.e. a real current sequence number.

The FastRTPS code dealing with gap messages was actually changed significantly in v1.9, to no longer use an std::set.  It now has a dedicated RTPSGapBuilder class, which does things more efficiently. Running the above blocking test with ROS2 foxy (v1.9.X) shows the blocking duration to not spike there. So perhaps this can be considered fixed in 1.9. Though it still looks to be iterating through all the way from 0, so the below fix might still be an improvement in 1.9.

## Fix
```
diff --git a/src/cpp/rtps/writer/ReaderProxy.cpp b/src/cpp/rtps/writer/ReaderProxy.cpp
index ba6d8635..13c856e7 100644
--- a/src/cpp/rtps/writer/ReaderProxy.cpp
+++ b/src/cpp/rtps/writer/ReaderProxy.cpp
@@ -70,7 +70,23 @@ void ReaderProxy::start(const RemoteReaderAttributes& reader_attributes)
     nack_supression_event_->reader_guid(reader_attributes_.guid);
     timers_enabled_.store(reader_attributes_.endpoint.reliabilityKind == RELIABLE);

     logInfo(RTPS_WRITER, "Reader Proxy started");
+    // The remote reader shouldn't care about sequence ids prior to the first
+    // one it receives.
+    SequenceNumber_t min_seq = writer_->get_seq_num_min();
+    if (min_seq != c_SequenceNumber_Unknown)
+        changes_low_mark_ = min_seq - 1;
+    else
+        changes_low_mark_ = writer_->next_sequence_number() - 2;
+    changes_low_mark_ = std::max(changes_low_mark_, SequenceNumber_t());
 }
```
It seems to work, the blocking duration no longer spikes. Though it turns out that the writer history is empty in many cases (get_seq_num_min is unknown), so we have to use next_sequence_number() as above.

Again, I'm not 100% sure if this is compliant with the RTPS spec.

**Comment by EduPonz** [comment]
Hi @aquad , and lots of thanks for your thorough description of the issue!

We were able to reproduce your issue in 1.8.2 (current ROS 2 Dashing). Even though the performance has significantly improved since 1.8.x, the issue will raise anyways if a later joiner appears not minutes but hours in the publication (meaning millions of changes).

For this reason, we have made a fix for master based on your idea (see #1045). What we've done is to simulate an the arrival of an initial acknack on `ReaderProxy::start()`, so that the ReaderProxy's low mark is set correctly. We are currently studying the migration of the fix to previous versions.

Again, thanks a lot!
